File: perl/mergesort.pl
use strict;
use warnings;

sub merge {
  my($s, @b) = @_;
  my(@a) = splice(@b, $s);

  if (@a * @b == 0) { return @a, @b; }
  my($head) = $a[0] < $b[0] ? shift(@a) : shift(@b);
  return $head, merge(int(@a), @a, @b);
}

sub mergesort {
  my($half) = int(@_ / 2);
  if ($half == 0) { return @_; }

  return merge($half, mergesort(splice(@_, $half)), mergesort(@_));
}

my(@sorted) = mergesort(10, 9, 8, 4, 5, 6, 7, 3, 2, 1);
print(join(" ", @sorted) . "\n");

File: haskell/mergesort.hs
merge a [] = a
merge [] b = b
merge as@(ah:at) bs@(bh:bt) =
	if ah <= bh
	then ah:(merge at bs)
	else bh:(merge as bt)

mergesort list
  | half == 0 = list
  | otherwise =
      let (a, b) = splitAt half list in
        merge (mergesort a) (mergesort b)

  where half = div (length list) 2

main = do
	print $ mergesort [10, 9, 8, 4, 5, 6, 7, 3, 2, 1]
